File: 백준/Silver/1406. 에디터/에디터.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        String str = br.readLine();
        int n = Integer.parseInt(br.readLine());

        Stack<Character> leftStack = new Stack<>();
        Stack<Character> rightStack = new Stack<>();

        for (char ch : str.toCharArray()) {
            leftStack.push(ch);
        }

        for (int i = 0; i < n; i++) {
            String input = br.readLine();

            if (input.startsWith("P")) {
                char c = input.charAt(2);
                leftStack.push(c);
            } else if (input.equals("L")) {
                if (!leftStack.isEmpty()) {
                    rightStack.push(leftStack.pop());
                }
            } else if (input.equals("D")) {
                if (!rightStack.isEmpty()) {
                    leftStack.push(rightStack.pop());
                }
            } else if (input.equals("B")) {
                if (!leftStack.isEmpty()) {
                    leftStack.pop();
                }
            }
        }

        while (!leftStack.isEmpty()) {
            rightStack.push(leftStack.pop());
        }

        while (!rightStack.isEmpty()) {
            sb.append(rightStack.pop());
        }

        System.out.print(sb);
    }
}

File: 백준/Silver/1406. 에디터/에디터.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val str = readLine()
    val list = LinkedList<Char>()
    for (i in str) {
        list.add(i)
    }
    val itr = list.listIterator()
    while (itr.hasNext()) {
        itr.next()
    }
    val n = readLine().toInt()
    repeat(n) {
        val ch = readLine()
        when (ch[0]) {
            'L' -> {
                if (itr.hasPrevious()) {
                    itr.previous()
                }
            }

            'D' -> {
                if (itr.hasNext()) {
                    itr.next()
                }
            }

            'B' -> {
                if (itr.hasPrevious()) {
                    itr.previous()
                    itr.remove()
                }
            }

            'P' -> {
                itr.add(ch[2])
            }
        }
    }
    println(list.toCharArray())
}